### PR TITLE
Fix int32 overflow in cunn_SoftMaxForward for rows near INT32_MAX

### LINKED ISSUE #191565: [CUDA] illegal memory read on cunn_SoftMaxForward
### 🐛 Describe the bug

```python
import torch
import torch.nn.functional as F


torch.cuda.init()
input_tensor = torch.empty_strided(
    (2_147_483_647,),
    (0,),
    dtype=torch.float16,
    device="cuda",
)
F.log_softmax(input_tensor, dim=-1, dtype=torch.float32)
torch.cuda.synchronize()
```

```
========= COMPUTE-SANITIZER
========= Invalid __global__ read of size 2 bytes
=========     at void at::native::<unnamed>::cunn_SoftMaxForward<(int)8, c10::Half, float, float, at::native::<unnamed>::LogSoftMaxForwardEpilogue>(T4 *, const T2 *, int)+0x650
=========     by thread (384,0,0) in block (0,0,0)
=========     Address 0x7fe4a0000300 is out of bounds
=========     and is 4294966528 bytes before the nearest allocation at 0x7fe5a0000000 of size 4294967296 bytes
```

### Versions

Collecting environment information...

/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      39%
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.13.0+cu126
[pip3] triton==3.7.1
[conda] Could not collect

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia

### PR DESCRIPTION
**Pull request by AyaanFaisal21**
## Issue

Fixes #191565

## Summary

`cunn_SoftMaxForward` takes the row length as a 32-bit `int`, so at `dim_size = 2**31 - 1` the index math in `ilpReduce` and `WriteFpropResults{,Vectorized}` wraps negative and reads before the buffer: an illegal memory read. 

This fix ports the `int64_t` length plus `can_use_32bit_indexing` dual dispatch that `cunn_SoftMaxBackward` already uses further down the same file, so the 32-bit path is still taken whenever the length fits. Full root cause and the reason existing 64-bit tests miss this are in #191565.

Verified on an A10 (CUDA 13.0): the reproducer is clean under `compute-sanitizer` after the change, the added test fails on an unpatched build of the same commit, and `test_nn.py` is otherwise unchanged. Timings across 8 shapes (median-of-7, 3 runs per build) moved in both directions and never by more than 0.41%.

This leaves the dual-dispatch shape duplicated between the forward and backward kernels. Factoring it into a shared helper would mean touching the working backward path and widening the diff, so I left it out, but happy to do that here or as a follow-up.

## Checklist

- [x] Passes lint (`spin fixlint`)
- [x] Added/updated tests
- [ ] Updated documentation (if applicable)
- [x] Included benchmark results (for PRs impacting perf)

## BC-breaking?

No.

### CODE REVIEW

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 1]
```diff
@@ -650,14 +654,14 @@ WriteBpropResultsVectorized(
 /**
  * This will apply the Epilogue with non-vectorized reads & writes for the general case
  */
-template <int ILP, typename scalar_t, typename accum_t, typename outscalar_t, template<typename, typename, typename> class Epilogue>
+template <int ILP, typename scalar_t, typename accum_t, typename outscalar_t, template<typename, typename, typename> class Epilogue, typename index_t = int32_t>
 __device__ __forceinline__ void
 WriteFpropResults(
-             int classes,
+             index_t classes,
```
Comment (Skylion007) [inline aten/src/ATen/native/cuda/SoftMax.cu:660]: How are the numerics even sensible or stable for this many classes? Do any gradients flow through lol?

Comment (AyaanFaisal21) [reply aten/src/ATen/native/cuda/SoftMax.cu:660]: I chose not to TORCH_CHECK the size because backward already handles the same case with 64-bit indexing (#104270), so I followed that existing path for forward rather than introducing a new size restriction, and checked that it transfers.

accscalar_t is untouched here; only the addressing changes, so the arithmetic is identical across the 32/64-bit boundary. Testing agrees: the same adversarial input gives 6.6e-3 error at N=2^30 on the 32-bit path versus 7.4e-3 at N=2^31-1, the first size taking the 64-bit path. Reg/Smem/Gmem already take index_t; cunn_SoftMaxForward was the remaining CUDA-reachable path that didn't.

And yes on gradients: I measured 1.0 at the signal index versus exact 1 - 1/N, and -4.60e-10 elsewhere versus exact -4.66e-10 in bf16. fp16 flushes the latter because they're below its subnormal range.

It's certainly a niche size, but given the existing backward precedent and that the common arithmetic is unchanged, preserving support here seems preferable to adding a new error boundary.
